internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,198 @@
+	if bindPort != "" {
+		mlport, err := strconv.Atoi(bindPort)
+		if err != nil {
+			sl.l.Log("op", "startup", "error", "unable to parse ml-bindport", "msg", err)
+			return nil, err
+		}
+		mconfig.BindPort = mlport
+		mconfig.AdvertisePort = mlport
+	}
+	loggerout := gokitlog.NewStdlibAdapter(gokitlog.With(sl.l, "component", "MemberList"))
+	mconfig.Logger = golog.New(loggerout, "", golog.Lshortfile)
+	if secret != "" {
+		sha := sha256.New()
+		mconfig.SecretKey = sha.Sum([]byte(secret))[:16]
+	}
+	sl.eventCh = make(chan memberlist.NodeEvent, 16)
+	mconfig.Events = &memberlist.ChannelEventDelegate{Ch: sl.eventCh}
+	var err error
+	sl.mList, err = memberlist.Create(mconfig)
+	if err != nil {
+		sl.l.Log("op", "startup", "error", err, "msg", "failed to create memberlist")
+		return nil, err
+	}
+	return &sl, nil
+}
+
+// Start starts the needed goroutines
+func (sl *SpeakerList) Start(client *k8s.Client) {
+	sl.client = client
+
+	if sl.mList == nil {
+		return
+	}
+
+	go sl.watchEvents()
+	go sl.memberlistJoinLoop()
+}
+
+func (sl *SpeakerList) memberlistJoin() {
+	s := sl.k8sSpeakers.Load()
+	if s == nil {
+		return
+	}
+
+	mlSpeakers := map[string]bool{}
+	for _, n := range sl.mList.Members() {
+		mlSpeakers[n.Name] = true
+	}
+	k8sSpeakers := s.(map[string]string)
+	joinList := []string{}
+	for name, ip := range k8sSpeakers {
+		if _, ok := mlSpeakers[name]; !ok {
+			joinList = append(joinList, ip)
+		}
+	}
+	if len(joinList) > 0 {
+		n, err := sl.mList.Join(joinList)
+		sl.l.Log("op", "ml-join", "msg", "Memberlist join", "nb joigned", n, "error ?", err, "join list", joinList)
+	}
+}
+
+func (sl *SpeakerList) memberlistJoinLoop() {
+	ticker := time.NewTicker(30 * time.Second)
+	for {
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			sl.memberlistJoin()
+		}
+	}
+}
+
+// SetSpeakersK8S update the list of speakers from the K8S API point of view
+func (sl *SpeakerList) SetSpeakersK8S(eps *v1.Endpoints) {
+	newmap := map[string]string{}
+	for _, subset := range eps.Subsets {
+		for _, ep := range subset.Addresses {
+			newmap[*ep.NodeName] = ep.IP
+		}
+	}
+	oldmap := sl.k8sSpeakers.Load()
+	if oldmap == nil || reflect.DeepEqual(oldmap.(map[string]string), newmap) {
+		sl.l.Log("op", "setSpeakersK8S", "msg", "New K8S Speakers list set", "list", newmap)
+		sl.k8sSpeakers.Store(newmap)
+		sl.memberlistJoin()
+		sl.client.ForceSync()
+	}
+}
+
+// UsableSpeakers return a map of usable nodes or an error
+func (sl *SpeakerList) UsableSpeakers() (map[string]string, error) {
+	s := sl.k8sSpeakers.Load()
+	if s == nil {
+		return nil, errors.New("k8sSpeakersNotSet")
+	}
+	k8sSpeakers := s.(map[string]string)
+	if sl.mList == nil {
+		return k8sSpeakers, nil
+	}
+
+	members := sl.mList.Members()
+	if len(k8sSpeakers) > 1 && float64(len(members)) < float64(len(k8sSpeakers))/2 {
+		// we see less than half of the speakers that are healthy in K8S
+		// to be safe let's fail
+		return nil, errors.New("memberlistSplitBrain")
+	}
```

Comment:
If you shutdown more than half of your speakers not cleanly, it'll take 40s to recover, but I'll recover (if you can speak to the API of course)

Reply:
When you go from 3 to 2 you are fine, when you go from 2 to 1 you are also fine (ie it's instant), it's only when you go from 3 to 1 not cleanly (crash) that it takes 40s, so it's the best behavior IMO

Reply:
Without this if block, the rest of the change still work and make it better IMO. This particular block helps in the case split brain happens and we will have two masters in each of the group without this if block.

My argument is that:
1. It's less likely that the node can't reach other nodes but still can see ARP request for it.
2. It makes the system less available as I described. It's reasonable that two of the 3 nodes are down at the same time because they share the same infrastructure (switch/power/physical host). We can't let it down for 40s in that case. I work on vmware sphere and it's pretty common that there are only two ESXI hosts that runs all the VMs. A failed host can easily take away half of the cluster with the other half not responding traffic for 40s.You expect it to be N+1 (even though two nodes are on the same infra), but it's actually not high available at all.
3. The knowledge gap you have to fill with your end users. This is not what most people would expect naturally. It's too much for the end users to design failure domain properly based on this formula. 

Lastly I believe the fact that there are less than half of active members in this group does't make it fundamentally less trustworthy. 2 vs 1, 10 vs 9.  I would recommend to hold this particular optimization until there are enough use cases reported

Reply:
@rata is not a fan of the splitbrain detection either. I'll remove it but it means we need to gratuitous ARP patch for sure (I'll see if there is anything to improve to not spam too much)

Reply:
removed